x-pack/plugins/spaces/server/new_platform/spaces_service/spaces_service.ts

```diff
@@ -0,0 +1,138 @@
+  scopedClient(request: RequestFacade): SpacesClient;
+
+  getSpaceId(request: RequestFacade): string;
+
+  isInDefaultSpace(request: RequestFacade): boolean;
+}
+
+interface CacheEntry {
+  spaceId: string;
+  isInDefaultSpace: boolean;
+}
+
+interface SpacesServiceDeps {
+  http: HttpServiceSetup;
+  elasticsearch: ElasticsearchServiceSetup;
+  savedObjects: SavedObjectsService;
+  getSecurity: () => SecurityPlugin | undefined;
+  config$: Observable<SpacesConfig>;
+  spacesAuditLogger: any;
+}
+
+export class SpacesService {
+  private configSubscription$?: Subscription;
+
+  private readonly contextCache: WeakMap<any, CacheEntry> = new WeakMap();
+
+  constructor(private readonly log: Logger, private readonly serverBasePath: string) {
+    this.contextCache = new WeakMap();
+  }
+
+  public async setup({
+    http,
+    elasticsearch,
+    savedObjects,
+    getSecurity,
+    config$,
+    spacesAuditLogger,
+  }: SpacesServiceDeps): Promise<SpacesServiceSetup> {
+    let config: SpacesConfig = await config$.pipe(first()).toPromise();
+
+    this.configSubscription$ = config$.subscribe({
+      next: updatedConfig => {
+        config = updatedConfig;
+      },
+    });
+
+    const adminClient = await elasticsearch.adminClient$.pipe(first()).toPromise();
+    return {
+      getSpaceId: (request: RequestFacade) => {
+        if (!this.contextCache.has(request)) {
+          this.populateCache(http, request);
+        }
+
+        const { spaceId } = this.contextCache.get(request) as CacheEntry;
+        return spaceId;
+      },
+      isInDefaultSpace: (request: RequestFacade) => {
+        if (!this.contextCache.has(request)) {
+          this.populateCache(http, request);
+        }
+
+        return this.contextCache.get(request)!.isInDefaultSpace;
+      },
+      scopedClient: (request: RequestFacade) => {
+        const internalRepository = savedObjects.getSavedObjectsRepository(
+          adminClient.callAsInternalUser
+        );
+
+        const callCluster = (endpoint: string, ...args: any[]) =>
+          adminClient.asScoped(request).callAsCurrentUser(endpoint, ...args);
+
+        const callWithRequestRepository = savedObjects.getSavedObjectsRepository(callCluster);
+
+        const security = getSecurity();
+        const authorization = security ? security.authorization : null;
+
+        return new SpacesClient(
+          spacesAuditLogger,
+          (message: string) => {
+            this.log.debug(message);
+          },
+          authorization,
+          callWithRequestRepository,
+          config,
+          internalRepository,
+          request
+        );
+      },
+    };
+  }
+
+  public async stop() {
+    if (this.configSubscription$) {
+      this.configSubscription$.unsubscribe();
+      this.configSubscription$ = undefined;
+    }
+  }
+
+  private populateCache(http: HttpServiceSetup, request: RequestFacade) {
+    const isLegacyRequest = typeof (request as any).getBasePath === 'function';
```

Comment:
can be deleted. legacy server uses the same method under the hood that knows how to work with NP and LP requests https://github.com/elastic/kibana/pull/35429/files#diff-8a36aecfdbb5ec8f3384bbc2c87c8b22L28

Reply:
I added this in to allow reporting to work correctly. Reporting abuses the `callWithRequest` handler by [creating a fake request](https://github.com/legrego/kibana/blob/4088c9c1de3d15f5cb0f4ba81e44231bddcecf85/x-pack/plugins/reporting/export_types/csv/server/execute_job.js#L52) to execute its jobs. This fake request ends up getting consumed by the Spaces Service as part of the saved objects client wrapper.

I'm open to other ideas here, because I don't love this, but it's hopefully a temporary solution until reporting is migrated to the NP.

Reply:
that wasn't obvious, thanks! we can add explicit flag
```js
    const fakeRequest = {
      isFakeRequest: true,
      headers: decryptedHeaders,
      getBasePath: () => basePath || serverBasePath,
    };
```